File: shared/src/main/java/ru/privatenull/pnauth/security/TotpService.java
package ru.privatenull.pnauth.security;

import ru.privatenull.pnauth.storage.AuthRepository;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Locale;
import java.util.UUID;

public final class TotpService {
    private static final String BASE32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
    private static final SecureRandom RANDOM = new SecureRandom();
    private final AuthRepository repository;
    private final byte[] encryptionKey;

    public TotpService(AuthRepository repository, byte[] encryptionKey) {
        if (encryptionKey == null || encryptionKey.length != 32) {
            throw new IllegalArgumentException("TOTP encryption key must be 32 bytes");
        }
        this.repository = repository;
        this.encryptionKey = encryptionKey.clone();
    }

    public String generateSecret() {
        byte[] bytes = new byte[20];
        RANDOM.nextBytes(bytes);
        return encodeBase32(bytes);
    }

    public boolean verify(String secret, String code) {
        if (secret == null || code == null || !code.matches("\\d{6}")) {
            return false;
        }
        long time = System.currentTimeMillis() / 30_000L;
        for (long offset = -1; offset <= 1; offset++) {
            if (generateCode(secret, time + offset).equals(code)) {
                return true;
            }
        }
        return false;
    }

    public String provisioningUri(String issuer, String username, String secret) {
        String safeIssuer = issuer == null || issuer.isBlank() ? "pnAuth" : issuer;
        return "otpauth://totp/" + encode(safeIssuer + ":" + username)
                + "?secret=" + secret + "&issuer=" + encode(safeIssuer) + "&algorithm=SHA1&digits=6&period=30";
    }

    public List<String> generateRecoveryCodes(int amount) {
        List<String> codes = new ArrayList<>();
        for (int i = 0; i < amount; i++) {
            StringBuilder code = new StringBuilder();
            for (int part = 0; part < 4; part++) {
                if (part > 0) code.append('-');
                for (int character = 0; character < 4; character++) {
                    code.append("ABCDEFGHJKLMNPQRSTUVWXYZ23456789".charAt(RANDOM.nextInt(32)));
                }
            }
            codes.add(code.toString());
        }
        return codes;
    }

    public void saveRecoveryCodes(UUID uniqueId, List<String> codes) {
        repository.clearRecoveryCodes(uniqueId);
        codes.forEach(code -> repository.addRecoveryCode(uniqueId, hashRecoveryCode(code)));
    }

    public boolean consumeRecoveryCode(UUID uniqueId, String code) {
        return repository.consumeRecoveryCode(uniqueId, hashRecoveryCode(code));
    }

    public String encrypt(String secret) {
        try {
            byte[] iv = new byte[12];
            RANDOM.nextBytes(iv);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(encryptionKey, "AES"), new GCMParameterSpec(128, iv));
            byte[] encrypted = cipher.doFinal(secret.getBytes(StandardCharsets.UTF_8));
            byte[] result = new byte[iv.length + encrypted.length];
            System.arraycopy(iv, 0, result, 0, iv.length);
            System.arraycopy(encrypted, 0, result, iv.length, encrypted.length);
            return Base64.getEncoder().encodeToString(result);
        } catch (Exception exception) {
            throw new IllegalStateException("Could not encrypt TOTP secret", exception);
        }
    }

    public String decrypt(String encrypted) {
        try {
            byte[] data = Base64.getDecoder().decode(encrypted);
            byte[] iv = java.util.Arrays.copyOfRange(data, 0, 12);
            byte[] payload = java.util.Arrays.copyOfRange(data, 12, data.length);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(encryptionKey, "AES"), new GCMParameterSpec(128, iv));
            return new String(cipher.doFinal(payload), StandardCharsets.UTF_8);
        } catch (Exception exception) {
            throw new IllegalStateException("Could not decrypt TOTP secret", exception);
        }
    }

    private static String generateCode(String secret, long counter) {
        try {
            byte[] key = decodeBase32(secret);
            byte[] message = new byte[8];
            for (int i = 7; i >= 0; i--) {
                message[i] = (byte) counter;
                counter >>>= 8;
            }
            Mac mac = Mac.getInstance("HmacSHA1");
            mac.init(new SecretKeySpec(key, "HmacSHA1"));
            byte[] hash = mac.doFinal(message);
            int offset = hash[hash.length - 1] & 0x0f;
            int binary = ((hash[offset] & 0x7f) << 24)
                    | ((hash[offset + 1] & 0xff) << 16)
                    | ((hash[offset + 2] & 0xff) << 8)
                    | (hash[offset + 3] & 0xff);
            return String.format(Locale.ROOT, "%06d", binary % 1_000_000);
        } catch (Exception exception) {
            throw new IllegalStateException("Could not verify TOTP code", exception);
        }
    }

    private static String hashRecoveryCode(String code) {
        try {
            return Base64.getEncoder().encodeToString(MessageDigest.getInstance("SHA-256")
                    .digest(code.replace("-", "").toUpperCase(Locale.ROOT).getBytes(StandardCharsets.UTF_8)));
        } catch (Exception exception) {
            throw new IllegalStateException("Could not hash recovery code", exception);
        }
    }

    private static String encodeBase32(byte[] bytes) {
        StringBuilder result = new StringBuilder((bytes.length * 8 + 4) / 5);
        int buffer = 0;
        int bits = 0;
        for (byte value : bytes) {
            buffer = (buffer << 8) | (value & 0xff);
            bits += 8;
            while (bits >= 5) {
                bits -= 5;
                result.append(BASE32.charAt((buffer >>> bits) & 31));
            }
        }
        if (bits > 0) result.append(BASE32.charAt((buffer << (5 - bits)) & 31));
        return result.toString();
    }

    private static byte[] decodeBase32(String value) {
        String normalized = value.replace("=", "").toUpperCase(Locale.ROOT);
        byte[] result = new byte[normalized.length() * 5 / 8];
        int buffer = 0;
        int bits = 0;
        int index = 0;
        for (char character : normalized.toCharArray()) {
            int valueIndex = BASE32.indexOf(character);
            if (valueIndex < 0) throw new IllegalArgumentException("Invalid Base32 secret");
            buffer = (buffer << 5) | valueIndex;
            bits += 5;
            if (bits >= 8) {
                bits -= 8;
                result[index++] = (byte) ((buffer >>> bits) & 0xff);
            }
        }
        return result;
    }

    private static String encode(String value) {
        return java.net.URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/security/TotpService.kt
package ru.privatenull.pnauth.security

import ru.privatenull.pnauth.storage.AuthRepository
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Arrays
import java.util.Base64
import java.util.Locale
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class TotpService(
    private val repository: AuthRepository,
    encryptionKey: ByteArray?
) {
    private val encryptionKey: ByteArray

    init {
        require(!(encryptionKey == null || encryptionKey.size != 32)) {
            "TOTP encryption key must be 32 bytes"
        }
        this.encryptionKey = encryptionKey.clone()
    }

    fun generateSecret(): String {
        val bytes = ByteArray(20)
        RANDOM.nextBytes(bytes)
        return encodeBase32(bytes)
    }

    fun verify(secret: String?, code: String?): Boolean {
        if (secret == null || code == null || !code.matches(Regex("\\d{6}"))) {
            return false
        }
        val time = System.currentTimeMillis() / 30_000L
        for (offset in -1L..1L) {
            if (MessageDigest.isEqual(
                    generateCode(secret, time + offset).toByteArray(StandardCharsets.US_ASCII),
                    code.toByteArray(StandardCharsets.US_ASCII)
                )
            ) {
                return true
            }
        }
        return false
    }

    fun provisioningUri(issuer: String?, username: String, secret: String): String {
        val safeIssuer = if (issuer.isNullOrBlank()) "pnAuth" else issuer
        return "otpauth://totp/" + encode("$safeIssuer:$username") +
                "?secret=" + secret + "&issuer=" + encode(safeIssuer) + "&algorithm=SHA1&digits=6&period=30"
    }

    fun generateRecoveryCodes(amount: Int): List<String> {
        val codes = ArrayList<String>()
        for (i in 0 until amount) {
            val code = StringBuilder()
            for (part in 0 until 4) {
                if (part > 0) code.append('-')
                for (character in 0 until 4) {
                    code.append("ABCDEFGHJKLMNPQRSTUVWXYZ23456789"[RANDOM.nextInt(32)])
                }
            }
            codes.add(code.toString())
        }
        return codes
    }

    fun saveRecoveryCodes(uniqueId: UUID, codes: List<String>) {
        repository.clearRecoveryCodes(uniqueId)
        codes.forEach { code -> repository.addRecoveryCode(uniqueId, hashRecoveryCode(code)) }
    }

    fun replaceTotpData(uniqueId: UUID, encryptedSecret: String, codes: List<String>) {
        repository.replaceTotpData(uniqueId, encryptedSecret, codes.map { hashRecoveryCode(it) })
    }

    fun clearTotpData(uniqueId: UUID) {
        repository.clearTotpData(uniqueId)
    }

    fun consumeRecoveryCode(uniqueId: UUID, code: String): Boolean {
        return repository.consumeRecoveryCode(uniqueId, hashRecoveryCode(code))
    }

    fun encrypt(secret: String): String {
        try {
            val iv = ByteArray(12)
            RANDOM.nextBytes(iv)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), GCMParameterSpec(128, iv))
            val encrypted = cipher.doFinal(secret.toByteArray(StandardCharsets.UTF_8))
            val result = ByteArray(iv.size + encrypted.size)
            System.arraycopy(iv, 0, result, 0, iv.size)
            System.arraycopy(encrypted, 0, result, iv.size, encrypted.size)
            return Base64.getEncoder().encodeToString(result)
        } catch (exception: Exception) {
            throw IllegalStateException("Could not encrypt TOTP secret", exception)
        }
    }

    fun decrypt(encrypted: String): String {
        try {
            val data = Base64.getDecoder().decode(encrypted)
            val iv = Arrays.copyOfRange(data, 0, 12)
            val payload = Arrays.copyOfRange(data, 12, data.size)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(encryptionKey, "AES"), GCMParameterSpec(128, iv))
            return String(cipher.doFinal(payload), StandardCharsets.UTF_8)
        } catch (exception: Exception) {
            throw IllegalStateException("Could not decrypt TOTP secret", exception)
        }
    }

    companion object {
        private const val BASE32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        private val RANDOM = SecureRandom()

        @JvmStatic
        fun hashRecoveryCode(code: String): String {
            try {
                return Base64.getEncoder().encodeToString(
                    MessageDigest.getInstance("SHA-256")
                        .digest(code.replace("-", "").uppercase(Locale.ROOT).toByteArray(StandardCharsets.UTF_8))
                )
            } catch (exception: Exception) {
                throw IllegalStateException("Could not hash recovery code", exception)
            }
        }

        private fun generateCode(secret: String, counter: Long): String {
            var c = counter
            try {
                val key = decodeBase32(secret)
                val message = ByteArray(8)
                for (i in 7 downTo 0) {
                    message[i] = c.toByte()
                    c = c ushr 8
                }
                val mac = Mac.getInstance("HmacSHA1")
                mac.init(SecretKeySpec(key, "HmacSHA1"))
                val hash = mac.doFinal(message)
                val offset = hash[hash.size - 1].toInt() and 0x0f
                val binary = ((hash[offset].toInt() and 0x7f) shl 24) or
                        ((hash[offset + 1].toInt() and 0xff) shl 16) or
                        ((hash[offset + 2].toInt() and 0xff) shl 8) or
                        (hash[offset + 3].toInt() and 0xff)
                return String.format(Locale.ROOT, "%06d", binary % 1_000_000)
            } catch (exception: Exception) {
                throw IllegalStateException("Could not verify TOTP code", exception)
            }
        }

        private fun encodeBase32(bytes: ByteArray): String {
            val result = StringBuilder((bytes.size * 8 + 4) / 5)
            var buffer = 0
            var bits = 0
            for (value in bytes) {
                buffer = (buffer shl 8) or (value.toInt() and 0xff)
                bits += 8
                while (bits >= 5) {
                    bits -= 5
                    result.append(BASE32[(buffer ushr bits) and 31])
                }
            }
            if (bits > 0) result.append(BASE32[(buffer shl (5 - bits)) and 31])
            return result.toString()
        }

        private fun decodeBase32(value: String): ByteArray {
            val normalized = value.replace("=", "").uppercase(Locale.ROOT)
            val result = ByteArray(normalized.length * 5 / 8)
            var buffer = 0
            var bits = 0
            var index = 0
            for (character in normalized.toCharArray()) {
                val valueIndex = BASE32.indexOf(character)
                require(valueIndex >= 0) { "Invalid Base32 secret" }
                buffer = (buffer shl 5) or valueIndex
                bits += 5
                if (bits >= 8) {
                    bits -= 8
                    result[index++] = ((buffer ushr bits) and 0xff).toByte()
                }
            }
            return result
        }

        private fun encode(value: String): String {
            return URLEncoder.encode(value, StandardCharsets.UTF_8)
        }
    }
}
